src/probe/probe.c

```diff
@@ -15,42 +22,529 @@
 #define trace_probe_error(__e, ...) \
 	trace_error(TRACE_CLASS_PROBE, __e, ##__VA_ARGS__)
 
+#define PROBE_DMA_INVALID	0xFFFFFFFF
+#define PROBE_POINT_INVALID	0xFFFFFFFF
+
+#define PROBE_BUFFER_LOCAL_SIZE		8192
+
+struct probe_dma_buf {
+	uintptr_t w_ptr;
+	uintptr_t r_ptr;
+	uintptr_t addr;
+	uintptr_t end_addr;
+	uint32_t size;
+	uint32_t avail;
+};
+
+struct probe_dma_ext {
+	uint32_t stream_tag;
+	uint32_t dma_buffer_size;
+	struct dma_sg_config config;
+	struct probe_dma_buf dmapb;
+	struct dma_copy dc;
+};
+
+struct probe_pdata {
+	struct probe_dma_ext ext_dma;
+	struct probe_dma_ext inject_dma[CONFIG_PROBE_DMA_MAX];
+	struct probe_point probe_points[CONFIG_PROBE_POINTS_MAX];
+};
+
+static struct probe_pdata *_probe;
+
+static int probe_dma_buffer_init(struct probe_dma_buf *buffer, uint32_t size)
+{
+	/* allocate new buffer */
+	buffer->addr = (uintptr_t)rballoc_align(SOF_MEM_ZONE_BUFFER,
+						SOF_MEM_CAPS_DMA, size, 128);
+
+	if (!buffer->addr) {
+		trace_probe_error("probe_dma_buffer_init() error: alloc failed");
+		return -ENOMEM;
+	}
+
+	bzero((void *)buffer->addr, size);
+	dcache_writeback_region((void *)buffer->addr, size);
+
+	/* initialise the DMA buffer */
+	buffer->size = size;
+	buffer->w_ptr = buffer->addr;
+	buffer->r_ptr = buffer->addr;
+	buffer->end_addr = buffer->addr + buffer->size;
+	buffer->avail = 0;
+
+	return 0;
+}
+
+static int probe_dma_init(struct probe_dma_ext *dma)
+{
+	struct dma_sg_config config;
+	uint32_t elem_size, elem_addr, elem_num;
+	int err = 0;
+
+	/* initialize dma buffer */
+	err = probe_dma_buffer_init(&dma->dmapb, PROBE_BUFFER_LOCAL_SIZE);
+	if (err < 0)
+		return err;
+
+	/* request HDA DMA in the dir LMEM->HMEM with shared access */
+	dma->dc.dmac = dma_get(DMA_DIR_LMEM_TO_HMEM, 0, DMA_DEV_HOST,
```

Comment:
This should work with generic DMA API and not just HDA DMA. i.e. it should work on CHT and BDW too, I guess the comment is wrong ?

Reply:
There are old comments related to this in dai.c and host.c that can also be fixed next refactor that touches them. (specifically mentioning HDA DMA instead of being generic)

Reply:
comment fixed